File: LeetCode/149MaxPointsOnALine/MaxPointsOnALine.cpp
class Solution {
public:
    int maxPoints(vector<vector<int>>& points) {
        // Compute slope -> Time:O(n^2), Space:O(n)
        map<double,int> slopeCount;
        double num, den, slope;
        int ans = 1;
        vector<int> currentPoint, nextPoint;
        for (int i = 0; i < points.size(); i++) {
            slopeCount.clear();
            currentPoint = points[i];
            for (int j = i + 1; j < points.size(); j++) {
                nextPoint = points[j];
                num = currentPoint[1] - nextPoint[1];
                den = currentPoint[0] - nextPoint[0];
                if (den == 0) {
                    slope = DBL_MAX;
                }
                else {
                    slope = num/den;
                }
                slopeCount[slope]++;
                ans = max(slopeCount[slope] + 1, ans);
            }
        }
        return ans;
    }
};

File: LeetCode/149MaxPointsOnALine/MaxPointsOnALine.py
class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        # Compute slope -> Time:O(n^2), Space:O(n)
        max_p = 1
        for i in range(len(points)):
            d = {}
            curr_p = points[i]
            for j in range(i+1, len(points)):
                next_p = points[j]
                num = (curr_p[1] - next_p[1])
                den = (curr_p[0] - next_p[0])
                if den == 0:
                    slope = float('inf')
                else:
                    slope = num / den
                if slope not in d:
                    d[slope] = set()
                d[slope].add(tuple(points[i]))
                d[slope].add(tuple(points[j]))
                if len(d[slope]) > max_p:
                    max_p = len(d[slope])
        return max_p
